lumicks/pylake/detail/value.py

```diff
@@ -0,0 +1,162 @@
+import math
+
+
+class ValueMixin:
+    """Mixin class to emulate numeric types. Requires a `value` attribute on the subclass.
+
+    Follows the description for Emulating Numeric Types:
+    (https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types)
+
+    The returned type is always a primitive type, since the metadata stored in subclasses of this
+    mixin cannot be guaranteed to still be accurate after any operation on the value.
+
+    Note: cannot be used in combination with dataclass, which overrrides the  `==` operator.
+
+    Operators will work for `str` and `bool` types stored in the `value` attribute;
+    however full behavior/methods for these types are not tested or guaranteed to work as expected.
+    """
+
+    @staticmethod
+    def extract_value(method):
+        def wrapper(self, other):
+            if isinstance(other, ValueMixin):
+                other = other.value
+
+            return method(self, other)
+
+        return wrapper
+
+    @extract_value
+    def __add__(self, other):
+        return self.value + other
+
+    @extract_value
+    def __sub__(self, other):
+        return self.value - other
+
+    @extract_value
+    def __mul__(self, other):
+        return self.value * other
+
+    @extract_value
+    def __truediv__(self, other):
+        return self.value / other
+
+    @extract_value
+    def __floordiv__(self, other):
+        return self.value // other
+
+    @extract_value
+    def __mod__(self, other):
+        return self.value % other
+
+    @extract_value
+    def __divmod__(self, other):
+        return divmod(self.value, other)
+
+    @extract_value
+    def __pow__(self, other):
+        return self.value**other
+
+    @extract_value
+    def __and__(self, other):
+        return self.value & other
+
+    @extract_value
+    def __xor__(self, other):
+        return self.value ^ other
+
+    @extract_value
+    def __or__(self, other):
+        return self.value | other
+
+    @extract_value
+    def __radd__(self, other):
+        return other + self.value
+
+    @extract_value
+    def __rsub__(self, other):
+        return other - self.value
+
+    @extract_value
+    def __rmul__(self, other):
+        return other * self.value
+
+    @extract_value
+    def __rtruediv__(self, other):
+        return other / self.value
+
+    @extract_value
+    def __rfloordiv__(self, other):
+        return other // self.value
+
+    @extract_value
+    def __rmod__(self, other):
+        return other % self.value
+
+    @extract_value
+    def __rdivmod__(self, other):
+        return divmod(other, self.value)
+
+    @extract_value
+    def __rpow__(self, other):
+        return other**self.value
+
+    @extract_value
+    def __rand__(self, other):
+        return other & self.value
+
+    @extract_value
+    def __rxor__(self, other):
+        return other ^ self.value
+
+    @extract_value
+    def __ror__(self, other):
+        return other | self.value
+
```

Comment:
I think you're missing one that's pretty important for control flow here.

```suggestion
    def __bool__(self):
        return bool(self.value)
```